### PR TITLE
Enable nullability in CollectionCodeDomSerializer and child classes

### PR DESCRIPTION
**Pull request by halgab**
I think this is the last public serializer that wasn't annotated (counting open PRs)

## Proposed changes

- Enable nullability in `CollectionCodeDomSerializer`
- Enable nullability in `DataGridViewRowCollectionCodeDomSerializer`
- Enable nullability in `TableLayoutControlCollectionCodeDomSerializer`

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/9927)

### CODE REVIEW

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs [hunk 1]
```diff
@@ -398,20 +383,20 @@ protected virtual object SerializeCollection(IDesignerSerializationManager manag
...
             {
                 Trace(TraceLevel.Error, "Cannot serialize arrays with rank > 1.");
-                manager.ReportError(string.Format(SR.SerializerInvalidArrayRank, ((Array)array).Rank.ToString(CultureInfo.InvariantCulture)));
+                manager.ReportError(string.Format(SR.SerializerInvalidArrayRank, array.Rank.ToString(CultureInfo.InvariantCulture)));
             }
             else
             {
                 // For an array, we need an array create expression.  First, get the array type
-                Type elementType = targetType.GetElementType();
+                Type elementType = targetType.GetElementType()!;
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs:399]: How did you reason this is not nullable?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs [hunk 2]
```diff
@@ -493,27 +477,19 @@ private CodeArrayCreateExpression SerializeArray(IDesignerSerializationManager m
         {
             Trace(TraceLevel.Verbose, $"Elements: {valuesToSerialize.Count}");
             // Here we need to invoke Add once for each and every item in the collection. We can re-use the property reference and method reference, but we will need to recreate the invoke statement each time.
-            CodeMethodReferenceExpression methodRef = new CodeMethodReferenceExpression(targetExpression, "Add");
+            CodeMethodReferenceExpression methodRef = new CodeMethodReferenceExpression(targetExpression!, "Add");
 
             if (valuesToSerialize.Count > 0)
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs:482]: This nesting is getting a bit much. Let's invert this to help reduce that.

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs [hunk 3]
```diff
@@ -588,25 +563,16 @@ private CodeArrayCreateExpression SerializeArray(IDesignerSerializationManager m
...
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs:564]: Ditto here

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs [hunk 4]
```diff
@@ -617,38 +583,38 @@ private CodeArrayCreateExpression SerializeArray(IDesignerSerializationManager m
...
-                            newCtx = new ExpressionContext(ctx.Expression, elementType, ctx.Owner);
-                            manager.Context.Push(newCtx);
+                            newContext = new ExpressionContext(ctx.Expression, elementType, ctx.Owner);
+                            manager.Context.Push(newContext);
                         }
 
                         try
                         {
-                            exp = SerializeToExpression(manager, o);
+                            expression = SerializeToExpression(manager, o);
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs:598]: Thanks for fixing these names! :)

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs (hunk 1 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs:399]: I followed [the docs](https://learn.microsoft.com/en-us/dotnet/api/system.type.getelementtype?view=net-7.0&devlangs=csharp&f1url=%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(System.Type.GetElementType)%3Bk(SolutionItemsProject)%3Bk(DevLang-csharp)%26rd%3Dtrue) of `GetElementType`:

> [Returns] the [Type](https://learn.microsoft.com/en-us/dotnet/api/system.type?view=net-7.0) of the object encompassed or referred to by the current array, pointer, or reference type, or null if the current [Type](https://learn.microsoft.com/en-us/dotnet/api/system.type?view=net-7.0) is not an array or a pointer, or is not passed by reference, or represents a generic type or a type parameter in the definition of a generic type or generic method.

Here `targetType` is an array type so we should be fine

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs [hunk 5]
```diff
@@ -328,54 +312,55 @@ protected virtual object SerializeCollection(IDesignerSerializationManager manag
...
-                    parameters = method.GetParameters();
-                    if (parameters.Length == 1 && parameters[0].ParameterType.IsArray)
-                    {
-                        if (MethodSupportsSerialization(method))
+                    case "AddRange":
+                        ParameterInfo[] parameters = method.GetParameters();
+                        if (parameters is [{ ParameterType.IsArray: true }])
                         {
-                            addRangeMethods.Add(method);
+                            if (MethodSupportsSerialization(method))
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs:329]: nit: combine these if statements and ditto below

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs (hunk 1 above)
Comment (lonitra) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs:399]: I see, thank you!

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/TableLayoutControlCollectionCodeDomSerializer.cs [hunk 6]
```diff
@@ -30,28 +28,22 @@ protected override object SerializeCollection(IDesignerSerializationManager mana
...
             bool isTargetInherited = false;
-            ExpressionContext ctx = manager.Context[typeof(ExpressionContext)] as ExpressionContext;
 
-            if (ctx is not null && ctx.Expression == targetExpression)
+            if (manager.TryGetContext(out ExpressionContext? ctx) && ctx.Expression == targetExpression)
             {
-                IComponent comp = ctx.Owner as IComponent;
-
-                if (comp is not null)
+                if (ctx.Owner is IComponent comp)
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/TableLayoutControlCollectionCodeDomSerializer.cs:34]: Let's also combine these if statement as well

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs [hunk 7]
```diff
@@ -236,24 +218,29 @@ public override object Serialize(IDesignerSerializationManager manager, object v
...
                     if (parameter is not null)
                     {
-                        Type type = parameter.ParameterType.IsArray ? parameter.ParameterType.GetElementType() : parameter.ParameterType;
+                        Type? type = parameter.ParameterType;
+                        if (type.IsArray)
+                        {
+                            type = type.GetElementType();
+                        }
+
                         if (type is not null && type.IsAssignableFrom(objType))
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs:244]: If `typ` is not supposed to be nullable I don't think this null check is necessary.

**Review by lonitra — COMMENTED** [review]
Just a few small comments otherwise LGTM

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs [hunk 8]
```diff
@@ -236,24 +218,29 @@ public override object Serialize(IDesignerSerializationManager manager, object v
...
+            Type? candidateType = null;
             if (final is null || (finalType is not null && !finalType.IsAssignableFrom(objType)))
             {
                 foreach (MethodInfo method in methods)
                 {
                     ParameterInfo parameter = method.GetParameters()[0];
                     if (parameter is not null)
                     {
-                        Type type = parameter.ParameterType.IsArray ? parameter.ParameterType.GetElementType() : parameter.ParameterType;
+                        Type? type = parameter.ParameterType;
```
[the remark this answers is not in the document]
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CollectionCodeDomSerializer.cs:238]: Good catch

**Pull request merged**